File: app/src/main/java/org/jellyfin/androidtv/playback/overlay/actions/ClosedCaptionsAction.java
package org.jellyfin.androidtv.playback.overlay.actions;

import android.content.Context;
import android.view.Gravity;
import android.view.MenuItem;
import android.view.View;
import android.widget.PopupMenu;

import org.jellyfin.androidtv.R;
import org.jellyfin.androidtv.TvApp;
import org.jellyfin.androidtv.model.compat.SubtitleStreamInfo;
import org.jellyfin.androidtv.playback.PlaybackController;
import org.jellyfin.androidtv.playback.overlay.CustomPlaybackTransportControlGlue;
import org.jellyfin.androidtv.playback.overlay.LeanbackOverlayFragment;
import org.jellyfin.androidtv.util.Utils;

import java.util.List;

import timber.log.Timber;

public class ClosedCaptionsAction extends CustomAction {

    public ClosedCaptionsAction(Context context, CustomPlaybackTransportControlGlue customPlaybackTransportControlGlue) {
        super(context, customPlaybackTransportControlGlue);
        initializeWithIcon(R.drawable.ic_select_subtitle);
    }

    @Override
    public void handleClickAction(PlaybackController playbackController, LeanbackOverlayFragment leanbackOverlayFragment, Context context, View view) {
        if (playbackController.getCurrentStreamInfo() == null) {
            Timber.w("StreamInfo null trying to obtain subtitles");
            Utils.showToast(TvApp.getApplication(), "Unable to obtain subtitle info");
            return;
        }
        List<SubtitleStreamInfo> subtitles = playbackController.getSubtitleStreams();
        PopupMenu subMenu = Utils.createPopupMenu(context, view, Gravity.END);
        MenuItem none = subMenu.getMenu().add(0, -1, 0, context.getString(R.string.lbl_none));
        int currentSubIndex = playbackController.getSubtitleStreamIndex();
        if (currentSubIndex < 0) none.setChecked(true);
        for (SubtitleStreamInfo sub : subtitles) {
            MenuItem item = subMenu.getMenu().add(0, sub.getIndex(), sub.getIndex(), sub.getDisplayTitle());
            if (currentSubIndex == sub.getIndex()) item.setChecked(true);
        }
        subMenu.getMenu().setGroupCheckable(0, true, false);
        subMenu.setOnDismissListener(new PopupMenu.OnDismissListener() {
            @Override
            public void onDismiss(PopupMenu menu) {
                leanbackOverlayFragment.setFading(true);
            }
        });
        subMenu.setOnMenuItemClickListener(new PopupMenu.OnMenuItemClickListener() {
            @Override
            public boolean onMenuItemClick(MenuItem item) {
                playbackController.switchSubtitleStream(item.getItemId());
                return true;
            }
        });
        subMenu.show();
    }
}

File: app/src/main/java/org/jellyfin/androidtv/ui/playback/overlay/action/ClosedCaptionsAction.kt
package org.jellyfin.androidtv.ui.playback.overlay.action

import android.content.Context
import android.view.Gravity
import android.view.View
import android.widget.PopupMenu
import android.widget.Toast
import org.jellyfin.androidtv.R
import org.jellyfin.androidtv.ui.playback.PlaybackController
import org.jellyfin.androidtv.ui.playback.overlay.CustomPlaybackTransportControlGlue
import org.jellyfin.androidtv.ui.playback.overlay.VideoPlayerAdapter
import org.jellyfin.androidtv.ui.playback.setSubtitleIndex
import org.jellyfin.sdk.model.api.MediaStreamType
import timber.log.Timber

class ClosedCaptionsAction(
	context: Context,
	customPlaybackTransportControlGlue: CustomPlaybackTransportControlGlue,
) : CustomAction(context, customPlaybackTransportControlGlue) {
	companion object {
		private const val ITEM_SET_OFFSET = Int.MIN_VALUE
	}

	private var popup: PopupMenu? = null
	private val subtitleOffsetPopup = SubtitleOffsetPopup(context)

	init {
		initializeWithIcon(R.drawable.ic_select_subtitle)
	}

	override fun handleClickAction(
		playbackController: PlaybackController,
		videoPlayerAdapter: VideoPlayerAdapter,
		context: Context,
		view: View,
	) {
		subtitleOffsetPopup.dismiss()

		if (playbackController.currentStreamInfo == null) {
			Timber.w("StreamInfo null trying to obtain subtitles")
			Toast.makeText(context, "Unable to obtain subtitle info", Toast.LENGTH_LONG).show()
			return
		}

		videoPlayerAdapter.leanbackOverlayFragment.setFading(false)
		removePopup()
		var openingSubtitleOffsetPopup = false
		popup = PopupMenu(context, view, Gravity.END).apply {
			with(menu) {
				var order = 0

				if (videoPlayerAdapter.hasTimingAdjustableSubtitle()) {
					add(1, ITEM_SET_OFFSET, order++, context.getString(R.string.lbl_subtitle_offset))
				}

				add(0, -1, order++, context.getString(R.string.lbl_none)).apply {
					isChecked = playbackController.subtitleStreamIndex == -1
				}

				for (sub in playbackController.currentMediaSource.mediaStreams.orEmpty()) {
					if (sub.type != MediaStreamType.SUBTITLE) continue

					add(0, sub.index, order++, sub.displayTitle).apply {
						isChecked = sub.index == playbackController.subtitleStreamIndex
					}
				}

				setGroupCheckable(0, true, false)
			}
			setOnDismissListener {
				if (!openingSubtitleOffsetPopup) {
					videoPlayerAdapter.leanbackOverlayFragment.setFading(true)
				}
				popup = null
			}
			setOnMenuItemClickListener { item ->
				if (item.itemId == ITEM_SET_OFFSET) {
					openingSubtitleOffsetPopup = true
					view.post {
						subtitleOffsetPopup.show(playbackController, videoPlayerAdapter)
					}
					false
				} else {
					playbackController.setSubtitleIndex(item.itemId)
					true
				}
			}
		}
		popup?.show()
	}

	fun removePopup() {
		popup?.dismiss()
		subtitleOffsetPopup.dismiss()
	}
}
